### PR TITLE
CDPS-1044: Initial sync endpoint

### PR DESCRIPTION
**Pull request by danielburnley**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/controller/DistinguishingMarksSyncController.kt [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+@RestController
+@Tag(name = "Distinguishing marks")
+@Tag(
+  name = "Sync with NOMIS",
+  description = "Endpoints to keep the Prison Person database in sync with changes in the NOMIS database",
+)
+@RequestMapping("/sync/prisoners/{prisonerNumber}/distinguishing-marks", produces = [MediaType.APPLICATION_JSON_VALUE])
+class DistinguishingMarksSyncController {
+  @PostMapping
+  @ResponseStatus(HttpStatus.OK)
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/controller/DistinguishingMarksSyncController.kt:45]: Should be CREATED for a 201 response?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/dto/request/distinguishingmarks/DistinguishingMarkCreationSyncRequest.kt [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import io.swagger.v3.oas.annotations.media.Schema
+import org.jetbrains.annotations.NotNull
+import java.time.ZonedDateTime
+
+@Schema(description = "Request object for syncing a prisoner's distinguishing marks")
+@JsonInclude(NON_NULL)
+class DistinguishingMarkCreationSyncRequest(
+  @Schema(
+    description = "Type of distinguishing mark. `ReferenceDataCode.id`.",
+    example = "MARK_TYPE_SCAR",
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/dto/request/distinguishingmarks/DistinguishingMarkCreationSyncRequest.kt:14]: We'll need to discuss who maps between NOMIS/DPS reference codes - it tripped us up on other syncs. (probably the sync service as that's can hold knowledge of both systems...)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/controller/DistinguishingMarksSyncController.kt [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+      required = true,
+    )
+    @PathVariable
+    distinguishingMarkId: UUID,
+    @RequestPart
+    @Parameter(
+      description = "File part of the multipart request",
+      required = false,
+    )
+    file: MultipartFile?,
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/controller/DistinguishingMarksSyncController.kt:252]: There's one other column from the NOMIS table OFFENDER_IMAGES called IMAGE_SOURCE_CODE which I haven't mentioned before - sorry! It contains values WEB, FILE or LIVE... not sure if we need to consider this, do you have any research on it or similar?

Comment (danielburnley) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/controller/DistinguishingMarksSyncController.kt:252]: Oh yeah the source code is interesting - We've discussed _knowing_ about it, but not whether or not we want to keep it. I'm not sure it's even exposed outside of NOMIS (e.g. via the prison API), if it is we dont handle it at the moment so we could potentially look at dropping it? I'll have an ask

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/dto/request/distinguishingmarks/DistinguishingMarkCreationSyncRequest.kt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    example = "BODY_PART_HEAD",
+  )
+  @NotNull("Body part is required.")
+  val bodyPart: String,
+
+  @Schema(
+    description = "Whether or not the image for the distinguishing mark has been updated, to inform us if we should update our stored image.",
+    example = "true",
+  )
+  val imageUpdated: Boolean?,
```
[the remark this answers is not in the document]
Comment (danielburnley) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/dto/request/distinguishingmarks/DistinguishingMarkCreationSyncRequest.kt:30]: Oh yeah I can remove this one, it was leftover from the prior discussions aha. The separate endpoints solve this issue 😅

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/controller/DistinguishingMarksSyncController.kt (hunk 3 above)
Comment (danielburnley) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/controller/DistinguishingMarksSyncController.kt:253]: I've added it in for now, along with the other "applies from/applies to" information I was missing - we can always remove it in the future

**Review by davidatmoj — DISMISSED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/controller/DistinguishingMarksSyncController.kt [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+  fun syncUpdateDistinguishingMarkImage(
+    @PathVariable
+    prisonerNumber: String,
+    @PathVariable
+    distinguishingMarkId: String,
+    @PathVariable
+    imageId: String,
+    @RequestBody
+    @Validated
+    request: DistinguishingMarkImageUpdateSyncRequest,
```
[the remark this answers is not in the document]
Comment (danielburnley) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonperson/controller/DistinguishingMarksSyncController.kt:309]: You can't replace the images only create/delete so this doesn't have one for now

**Review by davidatmoj — APPROVED** [review]

**Review by davidatmoj — APPROVED** [review]

**Pull request merged**